independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java

```diff
@@ -0,0 +1,140 @@
+/**
+ *
+ */
+package io.quarkus.bootstrap.resolver.maven;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.graph.Dependency;
+import org.eclipse.aether.graph.DependencyNode;
+import org.eclipse.aether.resolution.ArtifactRequest;
+
+import io.quarkus.bootstrap.model.AppArtifactKey;
+
+public class BuildDependencyGraphVisitor {
+
+    private final Set<AppArtifactKey> appDeps;
+    private final StringBuilder buf;
+    private final Consumer<String> buildTreeConsumer;
+    private final List<Boolean> depth;
+
+    private DependencyNode deploymentNode;
+    private DependencyNode runtimeNode;
+    private Artifact runtimeArtifact;
+
+    private final List<DependencyNode> deploymentDepNodes = new ArrayList<>();
+    private final List<ArtifactRequest> requests = new ArrayList<>();
+
+
+    public BuildDependencyGraphVisitor(Set<AppArtifactKey> appDeps, Consumer<String> buildTreeConsumer) {
+        this.appDeps = appDeps;
+        this.buildTreeConsumer = buildTreeConsumer;
+        if(buildTreeConsumer == null) {
+            buf = null;
+            depth = null;
+        } else {
+            buf = new StringBuilder();
+            depth = new ArrayList<>();
+        }
+    }
+
+    public List<DependencyNode> getDeploymentNodes() {
+        return deploymentDepNodes;
+    }
+
+    public List<ArtifactRequest> getArtifactRequests() {
+        return requests;
+    }
+
+    public void visit(DependencyNode node) {
+        final Dependency dep = node.getDependency();
+        if(dep == null) {
+            return;
+        }
+        if(depth != null) {
+            buf.setLength(0);
+            if (!depth.isEmpty()) {
+                for (int i = 0; i < depth.size() - 1; ++i) {
+                    if (depth.get(i)) {
+                        buf.append("|  ");
+                    } else {
+                        buf.append("   ");
+                    }
+                }
+                if (depth.get(depth.size() - 1)) {
+                    buf.append("|- ");
+                } else {
+                    buf.append("\\- ");
+                }
+            }
+            buf.append(dep.getArtifact()).append(':').append(dep.getScope());
```

Comment:
Does it print the classifier and extension?

Reply:
Yes, it's the same format as in `dependency:tree`

Reply:
Good to know!